### PR TITLE
Decimal performance improvements

### PR DESCRIPTION
**Pull request by lrubasze**
A set of changes to improve performance for `decimal` and `precise_decimal` types.

New math module has been implemented: `bnum_integer`.
It is strongly inspired by `integer` module, but it is based on [`bnum`](https://crates.io/crates/bnum) which offers:
_Arbitrary precision, fixed-size signed and unsigned integer types for Rust._

General performance improvement is  **95%**
(Detailed results to be provided.)

### CODE REVIEW

**Comment by lrubasze** [comment]
Performance measurements:


Operation | Decimal before | Decimal after | Gain | PreciseDecimal before | PreciseDecimal after | Gain
-- | -- | -- | -- | -- | -- | --
addition | 128.97 ns | 496.31 ps | -99.62% | 192.51 ns | 4.0719 ns | -97.88%
substraction | 99.590 ns | 476.98 ps | -99.52% | 203.51 ns | 3.1508 ns | -98.45%
multiplication | 1.5958 µs | 112.84 ns | -92.93% | 2.3290 µs | 668.78 ns | -71.28%
division | 1.4674 µs | 116.49 ns | -92.06% | 2.3264 µs | 646.91 ns | -72.19%
root | 8.7675 µs | 9.4003 µs | 7.22% | 84.682 µs | 85.380 µs | 0.82%
exponentiation | 7.1623 µs | 507.29 ns | -92.92% | 13.027 µs | 3.2570 µs | -75.00%
conversion from_string | 50.790 µs | 315.38 ns | -99.38% | 150.88 µs | 883.69 ns | -99.41%
conversion to_string | 58.378 µs | 541.70 ns | -99.07% | 240.39 µs | 1.4269 µs | -99.41%

File: radix-engine-interface/Cargo.toml [hunk 1]
```diff
@@ -18,6 +19,8 @@ paste = { version = "1.0.7"}
 serde = { version = "1.0.144", default-features = false, optional = true }
 serde_with = { version = "2.0.1", optional = true }
 strum = { version = "0.24", default-features = false, features = ["derive"] }
+# Use fork until fix for https://github.com/isaacholt100/bnum/issues/10 is not available
+bnum = { git = "https://github.com/lrubasze/bnum", branch = "do_not_override_features" , default-features = false, features = ["numtraits"] }
```
Comment (talekhinezh) [inline radix-engine-interface/Cargo.toml:23]: Can we use a `radixdlt` fork rather than a personal fork?

Comment (lrubasze) [reply radix-engine-interface/Cargo.toml]: that's a good point :)

**Review by iamyulong — APPROVED** [review]
LGTM

File: radix-engine-interface/src/math/decimal.rs [hunk 2]
```diff
@@ -450,91 +458,52 @@ impl FromStr for Decimal {
...
-        let mut a = self.0;
-        let mut buf = String::new();
-
-        let mut trailing_zeros = true;
-        for _ in 0..Self::SCALE {
-            let m: I256 = a % I256::from(10u8);
-            if m != 0.into() || !trailing_zeros {
-                trailing_zeros = false;
-                buf.push(char::from_digit(m.abs().to_u32().expect("Overflow"), 10).unwrap())
+        let multiplier = BnumI256::from(10i8).pow(Self::SCALE);
```
[the remark this answers is not in the document]
Comment (lrubasze) [reply radix-engine-interface/src/math/decimal.rs:488]: sure, good point

File: radix-engine-interface/src/math/bnum_integer.rs [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+                        Self(self.0.signum().try_into().unwrap())
+                    }
+
+                    /// Returns `true` if `self` is positive and `false` if the number is zero or
+                    /// negative.
+                    ///
+                    #[must_use]
+                    #[inline]
+                    pub fn is_positive(self) -> bool {
+                        self.0.is_positive().try_into().unwrap()
```
[the remark this answers is not in the document]
Comment (lrubasze) [reply radix-engine-interface/src/math/bnum_integer.rs:431]: copy&paste from integer.rs. thanks for spotting this.

File: radix-engine-interface/src/math/bnum_integer.rs [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+                    }
+
+                    /// Returns `true` if `self` is negative and `false` if the number is zero or
+                    /// positive.
+                    ///
+                    #[must_use]
+                    #[inline]
+                    pub fn is_negative(self) -> bool {
+                        self.0.is_negative().try_into().unwrap()
```
[the remark this answers is not in the document]
Comment (lrubasze) [reply radix-engine-interface/src/math/bnum_integer.rs:440]: Same

**Comment by lrubasze** [comment]
> LGTM, although I'm a bit concerned about the maturity of `bnum` (doens't seem to be widely adopted yet).

Thanks for review. Regarding maturity. I agree it is not widely adopted and there are other crates that are more popular.
The thing is they are not fully covering our requirements:
- signed integer types (decimal and precise_decimal is based on it)
- unsigned integer types (they are defined in integer.rs, however not sure if they are really needed..)
- scalable integer size
`bnum` offers that out of the box.

We could have taken more popular/mature crate and reimplement decimal/precise_decimal accordingly, but that would implicate higher regression risk, which I wanted to avoid.

**Pull request merged**